File: core/models/contact_schema.py
from sqlalchemy import Column, String, Boolean, DateTime, Index
from sqlalchemy.dialects.postgresql import UUID

from core.models.base import OrgScopedModel


class ContactSchema(OrgScopedModel):
    """An ORG-LEVEL, reusable contact-shape template (parent of ``SchemaField``).

    Replaces the old flat ``contact_field_definitions``: a schema groups a set of
    ``schema_fields`` and can be referenced by MANY directories (as a directory's
    ``default_schema_id`` — its manual-Add form + validation shape) and by MANY syncs
    (as the external→target mapping). It carries NO ``directory_id`` / ``datasource_id``
    / ``is_default`` — those bindings live on the referencing directory/sync.
    """

    __tablename__ = "contact_schemas"
    __table_args__ = (
        Index("ix_contact_schemas_org_name", "organization_id", "name"),
    )

    name = Column(String(120), nullable=False)
    description = Column(String(500), nullable=True)
    created_by_user_id = Column(UUID(as_uuid=True), nullable=True)
    is_active = Column(Boolean, nullable=False, default=True)
    deleted_at = Column(DateTime(timezone=True), nullable=True)

    def to_dict(self) -> dict:
        return {
            "id": str(self.id),
            "name": self.name,
            "description": self.description,
            "is_active": self.is_active,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: frontend/src/types/contactSchema.ts
/**
 * ContactSchema + SchemaField types — mirrors `core/models/contact_schema.py` and
 * `core/models/schema_field.py` `to_dict()`, plus the contact-schemas router shapes.
 *
 * A `ContactSchema` is an ORG-LEVEL, reusable contact-shape template that owns a set of
 * `SchemaField` rows. A directory references one via `default_schema_id` (its manual-Add
 * form + validation shape); a sync may pick any org schema for its external→target
 * mapping.
 */

/** Primitive field types the backend accepts (`_ALLOWED_FIELD_TYPES`). */
export type SchemaFieldType = 'string' | 'number' | 'integer' | 'boolean' | 'enum';

/**
 * The constrained JSON-Schema-subset validators the backend allows. String/enum fields
 * accept `minLength`/`maxLength`/`pattern`; number/integer accept `minimum`/`maximum`.
 */
export interface SchemaFieldValidators {
  minLength?: number;
  maxLength?: number;
  pattern?: string;
  minimum?: number;
  maximum?: number;
}

/** An enum choice for a `type: 'enum'` field. */
export interface SchemaFieldOption {
  value: string;
  label?: string;
}

/** A single field within a schema (mirrors `SchemaField.to_dict()`). */
export interface SchemaField {
  id: string;
  schema_id: string;
  field_name: string;
  label: string | null;
  is_mandatory: boolean;
  type: SchemaFieldType;
  format: string | null;
  validators: SchemaFieldValidators;
  options: SchemaFieldOption[] | null;
  /** External header this field reads from when used as a sync mapping. null = identity. */
  source_key: string | null;
  field_metadata: Record<string, unknown>;
  display_order: number;
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
}

/** A schema as returned by list/create/update (bare `to_dict`, no fields). */
export interface ContactSchema {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
}

/**
 * How many directories/syncs currently point at a schema — surfaced on
 * get-detail / update / delete so the UI can warn that a change is shared.
 */
export interface SchemaReferencedBy {
  directories: number;
  syncs: number;
}

/**
 * `GET /contact-schemas/{id}` — the schema enriched with its ACTIVE fields and the
 * `referenced_by` counts (from `get_schema_detail`).
 */
export interface ContactSchemaDetail extends ContactSchema {
  fields: SchemaField[];
  referenced_by: SchemaReferencedBy;
}

/**
 * `PATCH /contact-schemas/{id}` — the updated schema plus its `referenced_by` count so
 * the caller can warn the edit is shared.
 */
export interface ContactSchemaUpdateResult extends ContactSchema {
  referenced_by: SchemaReferencedBy;
}

/** `DELETE /contact-schemas/{id}` — soft-deleted; returns the reference counts it had. */
export interface ContactSchemaDeleteResult {
  ok: boolean;
  referenced_by: SchemaReferencedBy;
}

/** Body for `POST /contact-schemas` (create; admin/owner). */
export interface CreateSchemaPayload {
  name: string;
  description?: string | null;
  is_active?: boolean;
}

/** Body for `PATCH /contact-schemas/{id}` (update; admin/owner). */
export interface UpdateSchemaPayload {
  name?: string | null;
  description?: string | null;
  is_active?: boolean;
}

/** Body for `POST /contact-schemas/{id}/fields` (create field; admin/owner). */
export interface CreateSchemaFieldPayload {
  field_name: string;
  type?: SchemaFieldType;
  label?: string | null;
  format?: string | null;
  is_mandatory?: boolean;
  validators?: SchemaFieldValidators | null;
  options?: SchemaFieldOption[] | null;
  source_key?: string | null;
  display_order?: number;
  /** Per-field config; for date/datetime fields: `{ datetime_format, timezone }`. */
  field_metadata?: Record<string, unknown> | null;
}

/** Body for `PATCH /schema-fields/{id}` (update field; admin/owner). */
export interface UpdateSchemaFieldPayload {
  label?: string | null;
  type?: SchemaFieldType;
  format?: string | null;
  is_mandatory?: boolean;
  validators?: SchemaFieldValidators | null;
  options?: SchemaFieldOption[] | null;
  source_key?: string | null;
  display_order?: number;
  is_active?: boolean;
  field_metadata?: Record<string, unknown> | null;
}
